Cleanup fmod vs IEEE remainder

tautschnig opened the pull request:
1. Add tests for both the C and Java front-end.
2. Make sure we support both variants of the modulus operator on floating-point values in the back-ends: the SMT back-end only supported remainder (and only for FPA mode), the propositional back-end did not distinguish between fmod and IEEE remainder.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
Asking for help from @martin-cs: the remaining issue is that we need an ability to round-to-integer (`roundToIntegral` in SMT2) in our back-ends.

tautschnig committed:
Use FMA for IEEE remainder correction in float_utilst

Replace the extended-precision approach with FMA-based comparison for
the IEEE remainder correction step. FMA computes x - n*y with a single
rounding (the product n*y is exact internally), which provides a much
stronger soundness argument than the extended-precision approach.

Soundness proof sketch (included as comments in the code):

(a) The tentative n from round_to_integral(fp_div(x,y)) is off by at
    most 1 from the correct n_correct.

(b) fma(-n_correct, y, x) = R_correct exactly, because IEEE 754
    specifies that the remainder is exactly representable.

(c) fma(-n_wrong, y, x) = round(R_wrong). Since |R_wrong| >= |y|/2
    and |R_correct| <= |y|/2, the gap |R_wrong| - |R_correct| >= ulp(y),
    which exceeds the FMA rounding error of 1/2 ulp(R_wrong).

(d) In the true tie case (|R_correct| == |y|/2), both FMA results are
    exact and equal in magnitude. The rint-chosen even n is kept, which
    is correct per IEEE 754 tie-breaking.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Use FMA for IEEE remainder correction in float_utilst

Replace the extended-precision approach with FMA-based comparison for
the IEEE remainder correction step. FMA computes x - n*y with a single
rounding (the product n*y is exact internally), which provides a much
stronger soundness argument than the extended-precision approach.

Soundness proof sketch (included as comments in the code):

(a) The tentative n from round_to_integral(fp_div(x,y)) is off by at
    most 1 from the correct n_correct.

(b) fma(-n_correct, y, x) = R_correct exactly, because IEEE 754
    specifies that the remainder is exactly representable.

(c) fma(-n_wrong, y, x) = round(R_wrong). Since |R_wrong| >= |y|/2
    and |R_correct| <= |y|/2, the gap |R_wrong| - |R_correct| >= ulp(y),
    which exceeds the FMA rounding error of 1/2 ulp(R_wrong).

(d) In the true tie case (|R_correct| == |y|/2), both FMA results are
    exact and equal in magnitude. The rint-chosen even n is kept, which
    is correct per IEEE 754 tie-breaking.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Introduce ID_floatbv_mod to distinguish fmod from IEEE remainder

The existing ID_floatbv_rem was used for both fmod and IEEE remainder,
but these have different semantics (truncation vs round-to-nearest-even).
Add ID_floatbv_mod for fmod, keeping ID_floatbv_rem for IEEE remainder.

Update adjust_float_expressions to recognize the new ID, expr2c to
display it, and the incremental SMT2 back-end.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Introduce ID_floatbv_mod to distinguish fmod from IEEE remainder

The existing ID_floatbv_rem was used for both fmod and IEEE remainder,
but these have different semantics (truncation vs round-to-nearest-even).
Add ID_floatbv_mod for fmod, keeping ID_floatbv_rem for IEEE remainder.

Update adjust_float_expressions to recognize the new ID, expr2c to
display it, and the incremental SMT2 back-end.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Use __CPROVER_fmod and __CPROVER_remainder built-ins in C library

Replace the old __sort_of_CPROVER_remainder helper functions (which used
extended precision as a workaround) with direct calls to __CPROVER_fmod
and __CPROVER_remainder built-ins. These lower to ID_floatbv_mod and
ID_floatbv_rem respectively, which are handled natively by the solvers.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Use __CPROVER_fmod and __CPROVER_remainder built-ins in C library

Replace the old __sort_of_CPROVER_remainder helper functions (which used
extended precision as a workaround) with direct calls to __CPROVER_fmod
and __CPROVER_remainder built-ins. These lower to ID_floatbv_mod and
ID_floatbv_rem respectively, which are handled natively by the solvers.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Fix float_utilst::rem crash and distinguish fmod from remainder

The propositional back-end crashed on IEEE remainder because
to_signed_integer has a precondition requiring round-to-zero, but
remainder uses round-to-even.

Fix: use round_to_integral instead of to_signed_integer. The
round_to_integral function respects the current rounding mode
(ROUND_TO_ZERO for fmod, ROUND_TO_EVEN for IEEE remainder).

Set the rounding mode correctly in boolbv_floatbv_mod_rem based on
whether the expression is ID_floatbv_mod or ID_floatbv_rem.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Fix float_utilst::rem crash and distinguish fmod from remainder

The propositional back-end crashed on IEEE remainder because
to_signed_integer has a precondition requiring round-to-zero, but
remainder uses round-to-even.

Fix: use round_to_integral instead of to_signed_integer. The
round_to_integral function respects the current rounding mode
(ROUND_TO_ZERO for fmod, ROUND_TO_EVEN for IEEE remainder).

Set the rounding mode correctly in boolbv_floatbv_mod_rem based on
whether the expression is ID_floatbv_mod or ID_floatbv_rem.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add fmod and remainder support to SMT2 and float_bvt back-ends

SMT2 with FPA theory: add convert_floatbv_mod using fp.roundToIntegral
with roundTowardZero for fmod. IEEE remainder already used fp.rem.

SMT2 without FPA theory (float_bvt): add mod() and rem() methods that
compute x - round_to_integer(x/y) * y with appropriate rounding.

Use _Float16 in the regression test for faster SMT solving.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add fmod and remainder support to SMT2 and float_bvt back-ends

SMT2 with FPA theory: add convert_floatbv_mod using fp.roundToIntegral
with roundTowardZero for fmod. IEEE remainder already used fp.rem.

SMT2 without FPA theory (float_bvt): add mod() and rem() methods that
compute x - round_to_integer(x/y) * y with appropriate rounding.

Use _Float16 in the regression test for faster SMT solving.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add _Float16 variants of fmod/remainder/fma built-ins

Add __CPROVER_fmodf16, __CPROVER_remainderf16, __CPROVER_fmaf16
declarations and typechecker support. This enables verification of
floating-point remainder on _Float16, where the integer significand
division produces a manageable formula (~16K variables vs ~194K for
float).

Add a CORE regression test that verifies |remainder(x,y)| <= |y|/2
for ALL finite _Float16 inputs (completes in ~3s with MiniSat).

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add _Float16 variants of fmod/remainder/fma built-ins

Add __CPROVER_fmodf16, __CPROVER_remainderf16, __CPROVER_fmaf16
declarations and typechecker support. This enables verification of
floating-point remainder on _Float16, where the integer significand
division produces a manageable formula (~16K variables vs ~194K for
float).

Add a CORE regression test that verifies |remainder(x,y)| <= |y|/2
for ALL finite _Float16 inputs (completes in ~3s with MiniSat).

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening commented:
May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.

tautschnig committed:
Remove dead #if 0 block in boolbv_mod.cpp

Float-typed mod is handled by convert_floatbv_mod_rem, not here.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig commented:
> May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.

Will do, but I have a few more updates that are still being completed. Will create a separate PR once those are done.

tautschnig commented on src/util/floatbv_expr.h:
```diff
@@ -568,6 +568,71 @@ inline floatbv_rem_exprt &to_floatbv_rem_expr(exprt &expr)
+/// \brief Fused multiply-add expression: round(op0 * op1 + op2)
+/// with a single rounding. No explicit rounding mode operand;
+/// the solver uses the ambient rounding mode.
+class floatbv_fma_exprt : public multi_ary_exprt
```
Fixed. The comment now correctly states the rounding mode is stored as the 4th operand.

tautschnig commented on src/solvers/flattening/boolbv_floatbv_mod_rem.cpp:
```diff
@@ -19,7 +19,10 @@ bvt boolbvt::convert_floatbv_mod_rem(const binary_exprt &expr)
 
   float_utilst float_utils(prop);
 
-  auto rm = bv_utils.build_constant(ieee_floatt::ROUND_TO_EVEN, 32);
+  auto rm = bv_utils.build_constant(
+    expr.id() == ID_floatbv_rem ? ieee_floatt::ROUND_TO_EVEN
+                                : ieee_floatt::ROUND_TO_ZERO,
+    32);
   float_utils.set_rounding_mode(rm);
```
Updated the comment to reflect the current implementation.

tautschnig commented on src/solvers/floatbv/float_bv.cpp:
```diff
@@ -132,6 +132,22 @@ exprt float_bvt::convert(const exprt &expr) const
+    const auto &float_expr = to_binary_expr(expr);
+    return mod(float_expr.lhs(), float_expr.rhs());
+  }
+  else if(expr.id() == ID_floatbv_rem)
+  {
+    const auto &float_expr = to_binary_expr(expr);
+    return rem(float_expr.lhs(), float_expr.rhs());
+  }
+  else if(expr.id() == ID_floatbv_fma)
+  {
+    // float_bvt does not implement FMA; fall through to default handling
+    // which will report an error. A proper implementation would require
+    // double-width multiplication followed by addition and single rounding.
+  }
```
Implemented proper single-rounding FMA in `float_bvt`: multiply at double width (exact), align and add the third operand (exact), round once. This matches `float_utilst::fma` and ensures correctness of the remainder algorithm.

tautschnig commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -575,20 +684,106 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  bvt shift_dist = limit_distance(abs_exp_diff, mp_integer(int_width));
+
+  bvt mx = bv_utils.zero_extension(unpacked1.fraction, int_width);
+  bvt my = bv_utils.zero_extension(unpacked2.fraction, int_width);
```
Added a comment documenting that the width is O(2^e) bits, feasible for half/float/double but infeasible for long double/quad. The SMT FPA backend handles large formats efficiently.

tautschnig commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -575,20 +684,106 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+    prop.lor({unpacked1.zero, unpacked1.infinity, unpacked1.NaN});
+  fmod_result = bv_utils.select(x_is_special, src1, fmod_result);
+  fmod_result = bv_utils.select(unpacked2.infinity, src1, fmod_result);
+
```
Fixed. `fmod(±inf, y)`, `fmod(x, ±0)`, `fmod(x, NaN)`, and `fmod(NaN, y)` now correctly return NaN.

tautschnig committed:
Add proof verification CI and _Float16 property tests

Add a CI workflow that triggers when float implementation or proof
files change. It verifies:
1. Coq proofs compile (zero admits)
2. HOL Light proofs pass (zero mk_thm)
3. _Float16 exhaustive tests pass (linking proofs to implementation)

Add _Float16 tests for specific proved properties:
- special_cases: fmod(inf,y)=NaN, fmod(x,0)=NaN, etc.
- fmod_bound: |fmod(x,y)| < |y| for all finite inputs

Add cross-reference comments in float_utilst::rem mapping each
proof theorem to its corresponding _Float16 test.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add proof verification CI and _Float16 property tests

Add a CI workflow that triggers when float implementation or proof
files change. It verifies:
1. Coq proofs compile (zero admits)
2. HOL Light proofs pass (zero mk_thm)
3. _Float16 exhaustive tests pass (linking proofs to implementation)

Add _Float16 tests for specific proved properties:
- special_cases: fmod(inf,y)=NaN, fmod(x,0)=NaN, etc.
- fmod_bound: |fmod(x,y)| < |y| for all finite inputs

Add cross-reference comments in float_utilst::rem mapping each
proof theorem to its corresponding _Float16 test.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Update fmod/fmodf/fmodl regression tests

Verify concrete values and add float-overflow-check and nan-check.

Mark fmodl as THOROUGH due to long double divider size

The integer significand divider for long double (80-bit, e=15)
is 32834 bits wide, causing the test to take ~15 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Update fmod/fmodf/fmodl regression tests

Verify concrete values and add float-overflow-check and nan-check.

Mark fmodl as THOROUGH due to long double divider size

The integer significand divider for long double (80-bit, e=15)
is 32834 bits wide, causing the test to take ~15 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add Java fmod regression test (farith2)

Test the Java % operator on doubles, which uses fmod semantics.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add remainderf tie-break test

Test IEEE 754 tie-breaking behavior when quotient is near n+0.5.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Update remainder/remainderf/remainderl regression tests

Verify concrete values and add float-overflow-check and nan-check.

Mark remainderl as THOROUGH due to long double divider size

The integer significand divider for long double is 32834 bits wide,
causing the test to take ~4 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Update remainder/remainderf/remainderl regression tests

Verify concrete values and add float-overflow-check and nan-check.

Mark remainderl as THOROUGH due to long double divider size

The integer significand divider for long double is 32834 bits wide,
causing the test to take ~4 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add remainderf edge case test

Test remainder with inputs where fp division rounds the quotient
across an integer boundary.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add remainderf tie-break test

Test IEEE 754 tie-breaking behavior when quotient is near n+0.5.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add Java fmod regression test (farith2)

Test the Java % operator on doubles, which uses fmod semantics.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add remainderf edge case test

Test remainder with inputs where fp division rounds the quotient
across an integer boundary.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add HOL Light proofs for FMA-based remainder algorithm

Port core theorems using native int type with div/rem and
real_zpow for integer exponents. All proved with zero mk_thm.
Update check_proof.sh to verify both Coq and HOL Light.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add HOL Light proofs for FMA-based remainder algorithm

Port core theorems using native int type with div/rem and
real_zpow for integer exponents. All proved with zero mk_thm.
Update check_proof.sh to verify both Coq and HOL Light.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig commented:
> > May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.
> 
> Will do, but I have a few more updates that are still being completed. Will create a separate PR once those are done.

Now in https://github.com/diffblue/cbmc/pull/8884. The present PR should eventually be rebased on top of that.

tautschnig committed:
Add Coq proof and documentation for FMA-based remainder

Add a Coq/Flocq proof (doc/proofs/fma_remainder.v) that machine-checks
correctness properties of the FMA-based IEEE remainder algorithm.

Fully proved (zero admits, Coq 8.18 + Flocq):

  remainder_format: x - n*y is exactly representable in FLX format
  when |x - n*y| <= |y|/2 (both exponent orderings).

  fma_remainder_exact: round(x - n*y) = x - n*y, so FMA returns the
  exact remainder without rounding.

  comparison_step: the wrong candidate (n +/- 1) has strictly larger
  |remainder| than the correct one in exact arithmetic.

Add nearest_int_small, fmod_then_remainder, and
rounding_preserves_remainder_comparison. Move comparison_step
out of the FMA_Remainder section. Remove dead comments.
Add nearest_integer_from_ratio_lt for ey > ex case.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add Coq proof and documentation for FMA-based remainder

Add a Coq/Flocq proof (doc/proofs/fma_remainder.v) that machine-checks
correctness properties of the FMA-based IEEE remainder algorithm.

Fully proved (zero admits, Coq 8.18 + Flocq):

  remainder_format: x - n*y is exactly representable in FLX format
  when |x - n*y| <= |y|/2 (both exponent orderings).

  fma_remainder_exact: round(x - n*y) = x - n*y, so FMA returns the
  exact remainder without rounding.

  comparison_step: the wrong candidate (n +/- 1) has strictly larger
  |remainder| than the correct one in exact arithmetic.

Add nearest_int_small, fmod_then_remainder, and
rounding_preserves_remainder_comparison. Move comparison_step
out of the FMA_Remainder section. Remove dead comments.
Add nearest_integer_from_ratio_lt for ey > ex case.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig commented:
> Now in #8884. The present PR should eventually be rebased on top of that.

Rebase now done.

tautschnig committed:
Add fp.rem regression tests (Z3#2381, Z3#8414)

SMT-LIB tests:
- fp-rem-nonstandard/fp-rem1: fp.rem(3.0, 2.0) correctness (KNOWNBUG)
- fp-rem-nonstandard/fp-rem-nonstandard1: fp.rem on non-standard sort
- Tests from Z3 issues z3#2381 and z3#6553 (correctness, KNOWNBUG)

C test:
- Float-rem1: remainderf(3.0f, 2.0f) (KNOWNBUG)

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add fp.rem regression tests (Z3#2381, Z3#8414)

SMT-LIB tests:
- fp-rem-nonstandard/fp-rem1: fp.rem(3.0, 2.0) correctness (KNOWNBUG)
- fp-rem-nonstandard/fp-rem-nonstandard1: fp.rem on non-standard sort
- Tests from Z3 issues z3#2381 and z3#6553 (correctness, KNOWNBUG)

C test:
- Float-rem1: remainderf(3.0f, 2.0f) (KNOWNBUG)

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

martin-cs commented on regression/cbmc-library/fmod/main.c:
```diff
@@ -1,9 +1,19 @@
+  // https://stackoverflow.com/questions/2947044/how-do-i-use-modulus-for-float-double
+  // and
+  // https://stackoverflow.com/questions/25734144/difference-between-c-functions-remainder-and-fmod
+  double d1 = fmod(0.5, 0.3);
```
'0.3' as a decimal makes me nervous.

tautschnig replied:
Added comment explaining `0.3`/`0.2` are not exactly representable but results are correct on
nearest representable values. Also added hex float literal variants (`0x1.0p-1`, `0x1.3333333333333p-2`, `0x1.999999999999ap-3`).

martin-cs commented on regression/cbmc-library/remainderf/_Float16.c:
```diff
@@ -0,0 +1,30 @@
+{
+  _Float16 pos_inf_f16 = (_Float16)(1.0 / 0.0);
+  _Float16 x, y;
+  __CPROVER_assume(y != (_Float16)0.0);
```
Why ignore the NaN cases?

tautschnig replied:
Added NaN edge case tests: `remainder(NaN, y)`, `remainder(x, NaN)`, `remainder(x, 0)`, `remainder(inf, y)` - all assert the result is `NaN`. Added comment explaining the finite-input focus of the property test.

martin-cs commented on src/solvers/floatbv/float_bv.cpp:
```diff
@@ -953,6 +963,87 @@ exprt float_bvt::fma(
+  const constant_exprt round_to_zero =
+    from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet{2});
+
+  exprt div = convert(ieee_float_op_exprt{x, ID_floatbv_div, y, round_to_zero});
```
This feels like a different semantics to the one below.

tautschnig replied:
Added comment clarifying fmod uses truncation toward zero (C99 §7.12.10.1) while rem uses round-to-nearest-even (C99 §7.12.10.2).

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  bvt exp_diff = bv_utils.sub(exp1, exp2);
+  literalt ex_ge_ey = !exp_diff.back();
+  bvt abs_exp_diff = bv_utils.absolute_value(exp_diff);
+
```
Exponent difference seems correct and is important.

martin-cs replied:
Oh, also, you can use the sign bit to fast path because x < y => rem(x,y) = x in many cases.

tautschnig replied:
Thanks, implemented `|x| < |y| -> result = x` fast path. Added at the beginning (detection) and end (selection) of the function.

tautschnig replied:
Reverted as `|x| < |y|` does not imply `remainder(x,y) = x` when `|x| > |y|/2`. Concrete example: `remainder(0.7, 1.0)`. `x/y = 0.7`, nearest integer is `1`, so `remainder = 0.7 - 1.0 = -0.3`. The fast path would have incorrectly returned `0.7`.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  // Integer width for aligned significands.
+  // Note: this is O(2^e) bits, which is feasible for half (45 bits),
+  // float (282 bits), and double (2099 bits), but infeasible for
+  // long double/quad (32834 bits). Use the SMT FPA backend for those.
```
I'm not convinced there are any SMT solvers that will handle this better!  (Yet...)

Essentially this is the expensive bit of all approaches; you may need to do "exponent-difference" divide steps.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  // Note: this is O(2^e) bits, which is feasible for half (45 bits),
+  // float (282 bits), and double (2099 bits), but infeasible for
+  // long double/quad (32834 bits). Use the SMT FPA backend for those.
+  const std::size_t int_width = (std::size_t(1) << spec.e) + frac_bits + 2;
```
I'm not sure this is quite right.  For example for single precision you have 127 as the maximum exponent, -149 as the smallest and so the maximum difference is 276.  So if you want to avoid overflowing then you need 276+24 = 300 bits.  Maybe I am doing my calculations wrong.  Note that these cases are rare and generally simple in other ways.

tautschnig replied:
Yes, that's why this shouldn't be built without your oversight! Fixed to `(1 << e) + 2*p - 2`, which correctly accounts for `max_exp_diff = 2^e + p - 4` plus `p` significand bits. Updated comment with derivation.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  bvt mx = bv_utils.zero_extension(unpacked1.fraction, int_width);
+  bvt my = bv_utils.zero_extension(unpacked2.fraction, int_width);
+
+  // Align: shift the one with larger exponent left
```
Yes, hard agree!

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+    bv_utils.shift(my, bv_utilst::shiftt::SHIFT_LEFT, shift_dist));
+
+  // Integer remainder: fmod significand (unsigned)
+  bvt r_int = bv_utils.remainder(
```
Yes.  I would imagine this is most of the cost.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+    mx_aligned, my_aligned, bv_utilst::representationt::UNSIGNED);
+
+  // Integer quotient LSB (needed for remainder tie-breaking)
+  bvt q_int = bv_utils.divider(
```
IIRC you only need the LSB of the divider or something like that and you can do it by tapping the penultimate round of the remainder.  There might be something like doing one fewer divide step rounds and then doing the last one explicitly.

tautschnig replied:
Ack, but left this as a comment rather than implementing this for now.

tautschnig replied:
IMHO, we should independently look at the divider implementation and see whether that could be improved.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  fmod_unpacked.NaN = const_literal(false);
+  fmod_unpacked.infinity = const_literal(false);
+  fmod_unpacked.zero = bv_utils.is_zero(r_int);
+  bvt fmod_result = round_and_pack(fmod_unpacked);
```
The result should be exact so you shouldn't need to round.  I think with the architecture of the CProver encodings you may not have any choice though.

tautschnig replied:
Added comment explaining the fmod result is mathematically exact but `round_and_pack` is needed for normalization (subnormals, exponent adjustment).

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+  //   fmod(NaN, y)    = NaN
+  //   fmod(x, NaN)    = NaN
+  //   fmod(±0, y)     = ±0 (= x)
+  //   fmod(x, ±inf)   = x
```
From memory I think these are correct.  They should be listed in SymFPU thought.

tautschnig replied:
Reference added.

martin-cs commented on src/solvers/floatbv/float_utils.cpp:
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
+
+  if(!rounding_mode_bits.round_to_zero.is_true())
+  {
+    // Step 2: remainder(fmod, y) via FMA. |fmod/y| < 1, n ∈ {-1,0,1}.
```
I am not sure I see why this is needed.  I think you can get different rounding directions with a simple conditional subtract of y, ideally before you pack.

tautschnig replied:
Thank you, this makes a huge difference - see [3580205](https://github.com/diffblue/cbmc/pull/7885/commits/35802057a98556068da80ea19c2772062dca1fe9), including updated performance numbers.

martin-cs reviewed (commented):
I'm going to have to think about it a bit...

tautschnig committed:
Use conditional subtract for IEEE remainder correction

Replace the try-all-three FMA approach (n, n+1, n-1) with a simpler
conditional subtract: compare 2*|fmod| against |y| and correct by
±y when |fmod| > |y|/2, with tie-breaking by quotient parity.

This eliminates the div + round_to_integral + 3 FMA calls in
float_utils.cpp, and the n±1 multiply + subtract chain in
float_bv.cpp, replacing them with a single comparison and one
FMA (float_utils) or add/sub (float_bv).

Performance (_Float16 remainder property, exhaustive):
  Before: 4.6s, 85KB SMT2
  After:  0.4s, 66KB SMT2 (11x faster, 23% smaller)

Performance (float remainderf, fully symbolic):
  Before: >10 min timeout, 402K vars, 1.98M clauses, 86KB SMT2
  After:  17s,              383K vars, 1.89M clauses, 64KB SMT2

Re-enable the float remainderf benchmark (bench.desc) as CORE now
that it completes in ~17s.

Add Coq proofs (conditional_subtract_closer, no_subtract_when_small,
tie_case_equal_abs) and HOL Light proofs (CONDITIONAL_SUBTRACT_CLOSER,
NO_SUBTRACT_WHEN_SMALL, TIE_CASE_EQUAL_ABS) confirming the
conditional subtract gives the correct IEEE remainder in all cases.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Use conditional subtract for IEEE remainder correction

Replace the try-all-three FMA approach (n, n+1, n-1) with a simpler
conditional subtract: compare 2*|fmod| against |y| and correct by
±y when |fmod| > |y|/2, with tie-breaking by quotient parity.

This eliminates the div + round_to_integral + 3 FMA calls in
float_utils.cpp, and the n±1 multiply + subtract chain in
float_bv.cpp, replacing them with a single comparison and one
FMA (float_utils) or add/sub (float_bv).

Performance (_Float16 remainder property, exhaustive):
  Before: 4.6s, 85KB SMT2
  After:  0.4s, 66KB SMT2 (11x faster, 23% smaller)

Performance (float remainderf, fully symbolic):
  Before: >10 min timeout, 402K vars, 1.98M clauses, 86KB SMT2
  After:  17s,              383K vars, 1.89M clauses, 64KB SMT2

Re-enable the float remainderf benchmark (bench.desc) as CORE now
that it completes in ~17s.

Add Coq proofs (conditional_subtract_closer, no_subtract_when_small,
tie_case_equal_abs) and HOL Light proofs (CONDITIONAL_SUBTRACT_CLOSER,
NO_SUBTRACT_WHEN_SMALL, TIE_CASE_EQUAL_ABS) confirming the
conditional subtract gives the correct IEEE remainder in all cases.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -1659,9 +1659,9 @@ static smt_termt dispatch_expr_to_smt_conversion(
     return convert_expr_to_smt(*multiply, converted);
   }
 #if 0
-  else if(expr.id() == ID_floatbv_rem)
```
This change doesn't do anything.

tautschnig replied:
Cleaned and will be moved to a PR of its own.

kroening commented on regression/smt2_solver/fp-rem-nonstandard/fp-rem1.desc:
```diff
@@ -0,0 +1,9 @@
+^unsat$
+--
+Based on Z3#2381: fp.rem(3.0, 2.0) should be -1.0.
+CBMC's fp.rem appears to always return +0.0 (bug).
```
Is this now fixed?

tautschnig replied:
Cleaned up.

kroening commented on src/solvers/smt2/smt2_conv.h:
```diff
@@ -146,6 +146,7 @@ class smt2_convt : public stack_decision_proceduret
   void convert_floatbv_minus(const ieee_float_op_exprt &expr);
   void convert_floatbv_div(const ieee_float_op_exprt &expr);
   void convert_floatbv_mult(const ieee_float_op_exprt &expr);
+  void convert_floatbv_mod(const binary_exprt &expr);
```
floatbv_mod_exprt

tautschnig replied:
Done.

kroening commented on src/solvers/smt2/smt2_conv.h:
```diff
@@ -146,6 +146,7 @@ class smt2_convt : public stack_decision_proceduret
   void convert_floatbv_div(const ieee_float_op_exprt &expr);
   void convert_floatbv_mult(const ieee_float_op_exprt &expr);
+  void convert_floatbv_mod(const binary_exprt &expr);
   void convert_floatbv_rem(const binary_exprt &expr);
```
floatbv_rem_exprt

tautschnig replied:
Done.

tautschnig committed:
Fix float_bvt fmod/remainder for large exponent differences

float_bvt::mod/rem computed fmod as x - trunc(x/y)*y with the quotient
truncated to a fixed (type-width) integer.  When floor(x/y) exceeds that
width it wraps around, giving a wrong result -- e.g. fmod(2048, 2^-5)
(quotient 2^16) returned 2048 instead of 0, and the property
|fmod(x,y)| < |y| failed for _Float16 under the bit-vector SMT encoding.

Replace it with integer-significand arithmetic (align the significands,
take a single integer remainder), which is correct for arbitrarily large
exponent differences, mirroring float_utilst::rem.  mod and rem now share
a fmod_via_significand helper; rem keeps the conditional-subtract Step 2
for the IEEE round-to-nearest correction.

Adds regression/cbmc-library/remainderf/large_quotient.c demonstrating the
bug/fix on the default boolbv path, plus *-z3.desc tests exercising the
float_bvt bit-vector path via cbmc --z3.  large_quotient.c checks float
fmod at 2^33 (the smallest quotient that overflows the old 32-bit
truncation) and _Float16 fmod and remainder at 2^16: larger quotients, and
float remainderf's round-to-nearest tie-breaking, make the bit-vector
formula intractable for z3, so float remainderf's large-quotient behaviour
is left to the symbolic boolbv benchmark (bench.c).  The exhaustive
_Float16/fmod_bound z3 tests are tagged THOROUGH.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
